File: t/VB/DocumentCleanupDemo/UI/Command/DeskewDialog.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms

Imports Leadtools
Imports Leadtools.ImageProcessing
Imports Leadtools.ImageProcessing.Core

Namespace DocumentCleanupDemo
   Partial Public Class DeskewDialog : Inherits Form

      '' The DeskewCommand Class is part our LEAD Document Imaging funcions, this class will rotate the specified image to straighten it. This command is typically used to automatically straighten scanned images
      '' This dialog will use the DeskewCommandFlags, these flags will indicate whether to deskew the image, which background color to use, whether to deskew the image if the skew angle is very small, which type of interpolation to use, whether the image contains mostly text, and whether to use normal or fast rotation. 
      '' The Flags used are:
      '' DeskewCommandFlags.FillExposedArea, this flag is ued to fill areas exposed by rotation using the FillColor property
      '' DeskewCommandFlags.Threshold, this flag determines if you can deskew the image if the deskew angle is very small (less than 0.5 degrees).  
      '' DeskewCommandFlags.RotateLinear, this flag will not allow performing any interpolation methods when rotating 
      '' DeskewCommandFlags.DocumentImage, this flag will rotate the image. 

      Private _DeskewCommand As DeskewCommand = Nothing
      Public Sub New()
         InitializeComponent()
         _DeskewCommand = New DeskewCommand()
         InitializeUI()
      End Sub
      Public Sub New(ByVal DeskewCommand As DeskewCommand)
         InitializeComponent()
         _DeskewCommand = DeskewCommand
         InitializeUI()
      End Sub
      Public Property DeskewCommand() As DeskewCommand
         Get
            UpdateCommand()
            Return _DeskewCommand
         End Get
         Set(ByVal value As DeskewCommand)
            _DeskewCommand = value
            InitializeUI()
         End Set
      End Property

      Private Sub _btnOk_Click(ByVal sender As Object, ByVal e As EventArgs) Handles _btnOk.Click
         UpdateCommand()
         Me.DialogResult = Windows.Forms.DialogResult.OK
         Me.Close()
      End Sub

      Private Sub _btnCancel_Click(ByVal sender As Object, ByVal e As EventArgs) Handles _btnCancel.Click
         Me.DialogResult = Windows.Forms.DialogResult.Cancel
      End Sub

      Private Sub InitializeUI()
         '' Initialize the DeskewCommand Dialog with default values
         ' Fill
         If (_DeskewCommand.Flags And DeskewCommandFlags.FillExposedArea) = DeskewCommandFlags.FillExposedArea Then
            _cbFillExposedArea.Checked = True
            _lblColor.BackColor = Leadtools.Demos.Converters.ToGdiPlusColor(_DeskewCommand.FillColor)
            _lblColor.Enabled = True
         End If

         ' Threshold 
         If (_DeskewCommand.Flags And DeskewCommandFlags.Threshold) = DeskewCommandFlags.Threshold Then
            _cbThreshold.Checked = True
         End If

         ' Quality 
         If (_DeskewCommand.Flags And DeskewCommandFlags.RotateLinear) = DeskewCommandFlags.RotateLinear Then
            _rbLow.Checked = True
         End If
         If (_DeskewCommand.Flags And DeskewCommandFlags.RotateResample) = DeskewCommandFlags.RotateResample Then
            _rbMedium.Checked = True
         End If
         If (_DeskewCommand.Flags And DeskewCommandFlags.RotateBicubic) = DeskewCommandFlags.RotateBicubic Then
            _rbHigh.Checked = True
         End If

         ' Type
         If (_DeskewCommand.Flags And DeskewCommandFlags.DocumentImage) = DeskewCommandFlags.DocumentImage Then
            _rbTextOnly.Checked = True
         End If
         If (_DeskewCommand.Flags And DeskewCommandFlags.DocumentAndPictures) = DeskewCommandFlags.DocumentAndPictures Then
            _rbTextImages.Checked = True
         End If

         ' Speed
         If (_DeskewCommand.Flags And DeskewCommandFlags.UseNormalRotate) = DeskewCommandFlags.UseNormalRotate Then
            _rbNormal.Checked = True
         End If
         If (_DeskewCommand.Flags And DeskewCommandFlags.UseHighSpeedRotate) = DeskewCommandFlags.UseHighSpeedRotate Then
            _rbFast.Checked = True
         End If
      End Sub

      Private Sub UpdateCommand()
         '' Determine how the DeskewCommand will work by setting the values to its members and flags
         _DeskewCommand.Flags = CType(0, DeskewCommandFlags)

         ' Fill
         If _cbFillExposedArea.Checked Then
            _DeskewCommand.Flags = _DeskewCommand.Flags Or DeskewCommandFlags.FillExposedArea
            _DeskewCommand.FillColor = Leadtools.Demos.Converters.FromGdiPlusColor(_lblColor.BackColor)
         Else
            _DeskewCommand.Flags = _DeskewCommand.Flags Or DeskewCommandFlags.DoNotFillExposedArea
         End If

         ' Threshold
         If _cbThreshold.Checked Then
            _DeskewCommand.Flags = _DeskewCommand.Flags Or DeskewCommandFlags.Threshold
         Else
            _DeskewCommand.Flags = _DeskewCommand.Flags Or DeskewCommandFlags.NoThreshold
         End If

         ' Quality
         If _rbLow.Checked Then
            _DeskewCommand.Flags = _DeskewCommand.Flags Or DeskewCommandFlags.RotateLinear
         ElseIf _rbMedium.Checked Then
            _DeskewCommand.Flags = _DeskewCommand.Flags Or DeskewCommandFlags.RotateResample
         Else
            _DeskewCommand.Flags = _DeskewCommand.Flags Or DeskewCommandFlags.RotateBicubic
         End If

         ' Type
         If _rbTextOnly.Checked Then
            _DeskewCommand.Flags = _DeskewCommand.Flags Or DeskewCommandFlags.DocumentImage
         Else
            _DeskewCommand.Flags = _DeskewCommand.Flags Or DeskewCommandFlags.DocumentAndPictures
         End If

         ' Speed
         If _rbNormal.Checked Then
            _DeskewCommand.Flags = _DeskewCommand.Flags Or DeskewCommandFlags.UseNormalRotate
         Else
            _DeskewCommand.Flags = _DeskewCommand.Flags Or DeskewCommandFlags.UseHighSpeedRotate
         End If
      End Sub

      Private Sub _cbFillExposedArea_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles _cbFillExposedArea.CheckedChanged
         _lblColor.Enabled = _cbFillExposedArea.Checked
      End Sub

      Private Sub lblColor_Click(ByVal sender As Object, ByVal e As EventArgs) Handles _lblColor.Click
         Dim dlg As ColorDialog = New ColorDialog()
         dlg.Color = _lblColor.BackColor
         If dlg.ShowDialog() = Windows.Forms.DialogResult.OK Then
            _lblColor.BackColor = dlg.Color
         End If
      End Sub
   End Class
End Namespace

File: t/CS/DocumentCleanupDemo/UI/Command/DeskewDialog.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;

using Leadtools;
using Leadtools.ImageProcessing;
using Leadtools.ImageProcessing.Core;
using Leadtools.Demos;

namespace DocumentCleanupDemo
{
   public partial class DeskewDialog : Form
   {
      // The DeskewCommand Class is part our LEAD Document Imaging funcions, this class will rotate the specified image to straighten it. This command is typically used to automatically straighten scanned images
      // This dialog will use the DeskewCommandFlags, these flags will indicate whether to deskew the image, which background color to use, whether to deskew the image if the skew angle is very small, which type of interpolation to use, whether the image contains mostly text, and whether to use normal or fast rotation. 
      // The Flags used are:
      // DeskewCommandFlags.FillExposedArea, this flag is ued to fill areas exposed by rotation using the FillColor property
      // DeskewCommandFlags.Threshold, this flag determines if you can deskew the image if the deskew angle is very small (less than 0.5 degrees).  
      // DeskewCommandFlags.RotateLinear, this flag will not allow performing any interpolation methods when rotating 
      // DeskewCommandFlags.DocumentImage, this flag will rotate the image. 

      private DeskewCommand _DeskewCommand = null;
      public DeskewDialog()
      {
         InitializeComponent();
         _DeskewCommand = new DeskewCommand();
         InitializeUI();
      }
      public DeskewDialog(DeskewCommand DeskewCommand)
      {
         InitializeComponent();
         _DeskewCommand = DeskewCommand;
         InitializeUI();
      }
      public DeskewCommand DeskewCommand
      {
         get
         {
            UpdateCommand();
            return _DeskewCommand;
         }
         set
         {
            _DeskewCommand = value;
            InitializeUI();
         }
      }

      private void _btnOk_Click(object sender, EventArgs e)
      {
         UpdateCommand();
         this.DialogResult = DialogResult.OK;
         this.Close();
      }

      private void _btnCancel_Click(object sender, EventArgs e)
      {
         this.DialogResult = DialogResult.Cancel;
      }

      private void InitializeUI()
      {
         // Initialize the DeskewCommand Dialog with default values
         // Fill
         if ((_DeskewCommand.Flags & DeskewCommandFlags.FillExposedArea) == DeskewCommandFlags.FillExposedArea)
         {
            _cbFillExposedArea.Checked = true;
            _lblColor.BackColor = Converters.ToGdiPlusColor(_DeskewCommand.FillColor);
            _lblColor.Enabled = true;
         }

         // Threshold
         if ((_DeskewCommand.Flags & DeskewCommandFlags.Threshold) == DeskewCommandFlags.Threshold)
         {
            _cbThreshold.Checked = true;
         }

         // Quality
         if ((_DeskewCommand.Flags & DeskewCommandFlags.RotateLinear) == DeskewCommandFlags.RotateLinear)
         {
            _rbLow.Checked = true;
         }
         if ((_DeskewCommand.Flags & DeskewCommandFlags.RotateResample) == DeskewCommandFlags.RotateResample)
         {
            _rbMedium.Checked = true;
         }
         if ((_DeskewCommand.Flags & DeskewCommandFlags.RotateBicubic) == DeskewCommandFlags.RotateBicubic)
         {
            _rbHigh.Checked = true;
         }

         // Type
         if ((_DeskewCommand.Flags & DeskewCommandFlags.DocumentImage) == DeskewCommandFlags.DocumentImage)
         {
            _rbTextOnly.Checked = true;
         }
         if ((_DeskewCommand.Flags & DeskewCommandFlags.DocumentAndPictures) == DeskewCommandFlags.DocumentAndPictures)
         {
            _rbTextImages.Checked = true;
         }

         // Speed
         if ((_DeskewCommand.Flags & DeskewCommandFlags.UseNormalRotate) == DeskewCommandFlags.UseNormalRotate)
         {
            _rbNormal.Checked = true;
         }
         if ((_DeskewCommand.Flags & DeskewCommandFlags.UseHighSpeedRotate) == DeskewCommandFlags.UseHighSpeedRotate)
         {
            _rbFast.Checked = true;
         }
      }

      private void UpdateCommand()
      {
         // Determine how the DeskewCommand will work by setting the values to its members and flags
         _DeskewCommand.Flags = (DeskewCommandFlags)0;

         // Fill
         if (_cbFillExposedArea.Checked)
         {
            _DeskewCommand.Flags |= DeskewCommandFlags.FillExposedArea;
            _DeskewCommand.FillColor = Converters.FromGdiPlusColor(_lblColor.BackColor);
         }
         else
            _DeskewCommand.Flags |= DeskewCommandFlags.DoNotFillExposedArea;

         // Threshold
         if (_cbThreshold.Checked)
            _DeskewCommand.Flags |= DeskewCommandFlags.Threshold;
         else
            _DeskewCommand.Flags |= DeskewCommandFlags.NoThreshold;

         // Quality
         if (_rbLow.Checked)
         {
            _DeskewCommand.Flags |= DeskewCommandFlags.RotateLinear;
         }
         else if (_rbMedium.Checked)
         {
            _DeskewCommand.Flags |= DeskewCommandFlags.RotateResample;
         }
         else
         {
            _DeskewCommand.Flags |= DeskewCommandFlags.RotateBicubic;
         }

         // Type
         if (_rbTextOnly.Checked)
         {
            _DeskewCommand.Flags |= DeskewCommandFlags.DocumentImage;
         }
         else
         {
            _DeskewCommand.Flags |= DeskewCommandFlags.DocumentAndPictures;
         }

         // Speed
         if (_rbNormal.Checked)
         {
            _DeskewCommand.Flags |= DeskewCommandFlags.UseNormalRotate;
         }
         else
         {
            _DeskewCommand.Flags |= DeskewCommandFlags.UseHighSpeedRotate;
         }
      }

      private void _cbFillExposedArea_CheckedChanged(object sender, EventArgs e)
      {
         _lblColor.Enabled = _cbFillExposedArea.Checked;
      }

      private void lblColor_Click(object sender, EventArgs e)
      {
         ColorDialog dlg = new ColorDialog();
         dlg.Color = _lblColor.BackColor;
         if (dlg.ShowDialog() == DialogResult.OK)
         {
            _lblColor.BackColor = dlg.Color;
         }
      }
   }
}
